test_util/cluster_api.py

```diff
@@ -263,39 +276,76 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
         new_session.web_auth_default_user = user
         return new_session
 
-    def get_client(self, path, default_headers=None):
-        return test_util.helpers.ApiClient(
-            cluster=self,
-            user=self.web_auth_default_user,
+    def get_node_url(self, node, port=None):
+        """
+        Args:
+            node: (str) the hostname of the node to be requested from, if node=None,
+                then public cluster address (see environment DCOS_DNS_ADDRESS)
+            port: (int) port to be requested at. If port=None, the default port
+                for that given node type will be used
+        Returns:
+            fully-qualified URL string for this API
+        """
+        if node in self.masters:
+            role = 'master'
+        elif node in self.all_slaves:
+            role = 'agent'
+        else:
+            raise Exception('Node {} is not recognized within the DC/OS cluster'.format(node))
+        if port is None:
+            port = ADMINROUTER_PORT_MAPPING[role][self.scheme]
+        # do not explicitly declare default ports
+        if (port == 80 and self.scheme == 'http') or (port == 443 and self.scheme == 'https'):
+            netloc = node
+        else:
+            netloc = '{}:{}'.format(node, port)
+        return '{}://{}'.format(self.scheme, netloc)
+
+    def request_wrapper(self, request_fn, path, host_url=None, node=None, port=None, **kwargs):
+        """
+        Performs the same specialized request as ClusterApi but allows setting node
+        Args:
+            see ClusterApi for descriptions
+            node: host string corresponding to the ones stored in this object
+            port: if node is given, then a port can also be specified
+        """
+        if node is not None:
+            assert host_url is None, 'Cannot set both node and host_url'
+            host_url = self.get_node_url(node, port=port)
+        return super().request_wrapper(request_fn, path, host_url=host_url, **kwargs)
+
+    def get_client(self, path):
+        new_client = test_util.helpers.ApiClient(
+            default_host_url=self.dcos_url,
             api_base=path,
             ca_cert_path=self.ca_cert_path,
-            default_headers=default_headers)
+            default_headers=self.default_headers)
+        # give spawned clients the node= reqest option
+        new_client.request_wrapper = self.request_wrapper
 
     @property
     def marathon(self):
-        return test_util.marathon.Marathon(
-            cluster=self,
-            ca_cert_path=self.ca_cert_path,
-            user=self.web_auth_default_user)
+        marathon_client = test_util.marathon.Marathon(
+            default_host_url=self.dcos_url,
+            default_os_user=self.default_os_user,
+            default_headers=self.default_headers,
+            ca_cert_path=self.ca_cert_path)
+        # give spawned clients the node= request option
+        marathon_client.request_wrapper = self.request_wrapper
```

Comment:
@cmaloney having to manually attach a new wrapper is a little clunky, but it still seems better than getting another class